qemu/tests/win_virtio_driver_installer_pretest.py

```diff
@@ -0,0 +1,181 @@
+        time.sleep(2)
+        win_driver_installer_test.install_test_with_screen_on_desktop(
+            vm, session, test,
+            run_install_cmd,
+            installer_pkg_check_cmd,
+            copy_files_params=params)
+        error_context.context("Start viofs service...")
+        params["exe_find_cmd"] = 'dir /b /s VIOWIN_LTR\\EXE_FILE_NAME |' \
+                                 'findstr "\\EXE_MID_PATH\\\\"'
+        virtio_fs_utils.run_viofs_service(test, params, session)
+        params["exe_find_cmd"] = 'dir /b /s VIOWIN_LTR\\EXE_FILE_NAME'
+        change_virtio_media(params["cdrom_virtio"])
+        time.sleep(2)
+
+    # install/update virtio-win by installer
+    win_driver_installer_test.install_test_with_screen_on_desktop(
+        vm, session, test,
+        run_install_cmd,
+        installer_pkg_check_cmd,
+        copy_files_params=params)
+
+    if params.get("create_iso"):
+        guest_path = params.get("guest_path")
+        host_installer_file = params.get("host_installer_file")
+        host_iso_path = params.get("host_iso_path")
+        mkiso_cmd = params.get("mkiso_cmd")
+        process.run("rm -rf %s ; mkdir %s" % (host_iso_path, host_iso_path))
+        process.run("cp -r %s/* %s" % (host_installer_file, host_iso_path))
+        vm.copy_files_from(guest_path, host_iso_path)
+        process.run(mkiso_cmd)
+    else:
+        # some check after the installer is installed
+        win_driver_installer_test.win_installer_test(session, test, params)
+        win_driver_installer_test.check_gagent_version(
+            session, test,
+            gagent_pkg_info_cmd,
+            expected_gagent_version)
+        win_driver_installer_test.driver_check(session, test, params)
+        # balloon and virtiofs service check
+        fail_tests = []
+        for driver_name, device_name, device_hwid in zip(
+                win_driver_installer_test.driver_name_list,
+                win_driver_installer_test.device_name_list,
+                win_driver_installer_test.device_hwid_list):
+            if driver_name == "viofs" or driver_name == "balloon":
+                error_context.context("Check %s service." % driver_name,
+                                      test.log.info)
+                if driver_name == "viofs":
+                    output = virtio_fs_utils.query_viofs_service(test,
+                                                                 params,
+                                                                 session)
+                    if not re.search("stopped", output.lower(), re.M):
+                        fail_tests.append(driver_name)
+                else:
+                    balloon_test_win = BallooningTestWin(test, params, env)
+                    output = balloon_test_win.operate_balloon_service(
+                        session, "status")
+                    if not re.search("running", output.lower(), re.M):
+                        fail_tests.append(driver_name)
```

Comment:
The logic could be improved, I guess you want to avoid the logging line duplicated, but this way you went through one more `if` condition which produce less efficiency code, it hasn't to be 2 times same condition check.

Reply:
update.